rocky/views/upload_csv.py

```diff
@@ -63,24 +75,54 @@ def get_context_data(self, **kwargs):
         context["criterias"] = CSV_CRITERIAS
         return context
 
-    def get_or_create_network(self, network: str) -> Network:
-        if network in self.networks:
-            return self.networks[network]
-        network_ooi = Network(name=network)
-        self.networks[network] = network_ooi
-        return network_ooi
-
-    def get_ooi_from_csv(self, ooi_type: str, values: Dict[str, str]):
-        network = self.get_or_create_network(values.get("network", "internet"))
-        self._save_ooi(ooi=network, organization=self.organization_code)
-        if ooi_type == "Hostname":
-            return Hostname(name=values["name"], network=network.reference)
-        if ooi_type == "URL":
-            return URL(raw=values["raw"], network=network.reference)
-        if ooi_type == "IPAddressV4":
-            return IPAddressV4(address=values["address"], network=network.reference)
-        if ooi_type == "IPAddressV6":
-            return IPAddressV6(address=values["address"], network=network.reference)
+    def get_or_create_reference(self, ooi_type_name: str, value: str):
+        ooi_type_name = next(filter(lambda x: x.casefold() == ooi_type_name.casefold(), self.ooi_types.keys()))
+
+        # get from cache
+        cache = self.reference_cache.setdefault(ooi_type_name, {})
+        if value in cache:
+            return cache[value]
+
+        ooi_type = self.ooi_types[ooi_type_name]["type"]
+
+        # set default value if any
+        if value is None:
+            value = self.ooi_types[ooi_type_name].get("default")
+
+        # create the ooi
+        kwargs = {self.ooi_types[ooi_type_name]["argument"]: value}
+        ooi = ooi_type(**kwargs)
+        cache[value] = ooi
+
+        return ooi
+
+    def get_ooi_from_csv(self, ooi_type_name: str, values: Dict[str, str]):
+        ooi_type = self.ooi_types[ooi_type_name]["type"]
+        ooi_fields = [
+            (field, model_field.type_ == Reference, model_field.required)
+            for field, model_field in ooi_type.__fields__.items()
+            if field not in self.skip_properties
+        ]
+
+        kwargs = {}
+        for field, is_reference, required in ooi_fields:
+            if is_reference and required:
```

Comment:
What happens to referended, but not required values? These now end up as strings instead of their reference, right?

Reply:
Yes indeed, good catch

Reply:
But I'm not sure if they're relevant, since we only ask for the required fields

Reply:
I think it would be a good idea to have unit tests that cover the different scenarios

Reply:
We can ask for any field if we also support those optional fields, and I see no reason why we should limit the upload to only required fields or referenced fields.